File: week1/2798/2298.py
import sys
input = sys.stdin.readline

class Blackjack:
    def __init__(self):
        self.N, self.M = map(int, input().split())
        self.cards = list(map(int, input().split()))
        self.ans = -1

    def run(self):
        self.pick_cards(0, 0, 0)
        print(self.ans)

    def pick_cards(self, idx, picked, sum):
        if picked == 3:
            if sum <= self.M:
                self.ans = max(self.ans, sum)
            return

        if idx >= self.N or sum >= self.M:
            return
        
        # pick now
        self.pick_cards(idx+1, picked+1, sum+self.cards[idx])

        # don't pick
        self.pick_cards(idx+1, picked, sum)


if __name__ == "__main__":
    B = Blackjack()
    B.run()

File: week1/2798/2298.cpp
#include <iostream>
#include <vector>

using namespace std;

int N, M;
vector<int> cards;
int ans = 0;

void find_max_sum(int, int, int);

int main() {
    // set input
    cin >> N >> M;
    cards.reserve(N);
    int num;
    for (int i=0; i<N; i++) {
        cin >> num;
        cards.push_back(num);
    }

    // find answer
    find_max_sum(0, 0, 0);
    cout << ans;
}

void find_max_sum(int idx, int picked, int sum) {
    if (picked == 3) {
        if (sum <= M)
            ans = max(ans, sum);
        return;
    }
    
    if (idx >= N || sum >= M || ans == M) return;

    find_max_sum(idx+1, picked+1, sum+cards[idx]);
    find_max_sum(idx+1, picked, sum);
}